bin/runtest.ml

```diff
@@ -0,0 +1,141 @@
+        ; ( "Run tests sequentially without output buffering"
+          , "dune runtest --no-buffer -j 1" )
+        ]
+    ]
+  in
+  Cmd.info "runtest" ~doc ~man ~envs:Common.envs
+;;
+
+let find_cram_test path ~parent_dir =
+  let open Memo.O in
+  Source_tree.nearest_dir parent_dir
+  >>= Dune_rules.Cram_rules.cram_tests
+  (* We ignore the errors we get when searching for cram tests as they will
+     be reported during building anyway. We are only interested in the
+     presence of cram tests. *)
+  >>| List.filter_map ~f:Result.to_option
+  (* We search our list of known cram tests for the test we are looking
+     for. *)
+  >>| List.find ~f:(fun (test : Dune_rules.Cram_test.t) ->
+    let src =
+      match test with
+      | File src -> src
+      | Dir { dir = src; _ } -> src
+    in
+    Path.Source.equal path src)
+;;
+
+(** [disambiguate_test_name path] is a memoized function that takes in a
+    directory [path] and classifies it as either a cram test or a directory to
+    run tests in. *)
+let disambiguate_test_name path =
+  match Path.Source.parent path with
+  | None -> Memo.return @@ `Runtest (Path.source Path.Source.root)
+  | Some parent_dir ->
+    let open Memo.O in
+    find_cram_test path ~parent_dir
+    >>= (function
+     (* If we find the cram test, then we request that is run. *)
+     | Some test ->
+       Memo.return (`Test (Dune_rules.Cram_test.name test))
+       (* If we don't find it, then we assume the user intended a directory for
+          @runtest to be used. *)
+     | None ->
+       Source_tree.find_dir path
+       >>= (function
+        (* We need to make sure that this directory or file exists. *)
+        | Some _ -> Memo.return (`Runtest (Path.source path))
+        | None ->
+          (* If the user misspelled the test name, we give them a hint. *)
+          let+ hints =
+            let+ candidates =
+              let+ files = Source_tree.files_of parent_dir in
+              Path.Source.Set.to_list_map files ~f:Path.Source.basename
+            in
+            User_message.did_you_mean (Path.Source.basename path) ~candidates
+          in
+          User_error.raise
+            ~hints
+            [ Pp.textf "%S was not found." (Path.Source.to_string path) ]))
+;;
+
+let term =
+  let name = Arg.info [] ~docv:"TEST" in
+  let+ builder = Common.Builder.term
+  and+ dirs = Arg.(value & pos_all string [ "." ] name) in
+  let common, config = Common.init builder in
+  let request (setup : Import.Main.build_system) =
+    List.map dirs ~f:(fun dir ->
+      let open Action_builder.O in
+      let* alias_kind =
+        (* We always interpret the arguments as paths *)
+        let dir = Path.of_string dir in
+        match Path.as_in_source_tree dir with
+        (* If the path is in the source tree, we disambiguate it. *)
+        | Some path -> Action_builder.of_memo (disambiguate_test_name path)
+        (* Otherwise, we raise an error for the user. *)
+        | None ->
```

Comment:
Does this just forbid _build paths from being passed in runtest? Why should we do that?